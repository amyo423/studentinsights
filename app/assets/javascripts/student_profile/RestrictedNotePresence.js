import React from 'react';
import PropTypes from 'prop-types';
import {apiFetchJson} from '../helpers/apiFetchJson';
import NoteText from '../components/NoteText';
import GenericLoader from '../components/GenericLoader';


// Renders the presence of a restricted note, and with a click or action,
// allows viewing or editing.
export default class RestrictedNotePresence extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isViewing: false
    };

    this.fetchRestrictedContent = this.fetchRestrictedContent.bind(this);
    this.onViewClicked = this.onViewClicked.bind(this);
    this.renderRestrictedContent = this.renderRestrictedContent.bind(this);
  }

  fetchRestrictedContent() {
    const {eventNoteId} = this.props;
    return apiFetchJson(`/api/event_notes/${eventNoteId}/restricted_note_json`);
  }

  onViewClicked(e) {
    e.preventDefault();
    this.setState({isViewing: true});
  }

  render() {
    const {isViewing, allowViewing} = this.state;
    return (
      <div className="RestrictedNotePresence">
        {isViewing && !allowViewing
          ? this.renderViewing()
          : this.renderRedaction()}
      </div>
    );
  }

  renderRedaction() {
    const {studentFirstName, educatorName, allowViewing} = this.props;
    const studentFirstNameOrTheir = (studentFirstName)
      ? `${studentFirstName}'s`
      : 'their';

    return (
      <div>
        <NoteText
          style={styles.restrictedNoteRedaction}
          text={`To respect ${studentFirstNameOrTheir} privacy, ${educatorName || 'the author'} marked this note as restricted.`}
        />
        {allowViewing && 
          <a
            href="#"
            style={styles.showLink}
            onClick={this.onViewClicked}>show restricted note</a>
        }
      </div>
    );
  }

  renderViewing() {
    return (
      <GenericLoader
        promiseFn={this.fetchRestrictedContent}
        render={this.renderRestrictedContent}
      />
    );
  }

  renderRestrictedContent(json) {
    const {text} = json;
    return (
      <div>
        <NoteText text={text} />
        <div style={styles.restrictedNoteLabel}>This is a restricted note.</div>
      </div>
    );
  }
}
RestrictedNotePresence.propTypes = {
  eventNoteId: PropTypes.number.isRequired,
  studentFirstName: PropTypes.string,
  educatorName: PropTypes.string,
  allowViewing: PropTypes.bool
};
RestrictedNotePresence.defaultProps = {
  allowViewing: false
};

const styles = {
  restrictedNoteRedaction: {
    color: '#999'
  },
  showLink: {
    display: 'inline-block',
    color: '#999',
    cursor: 'pointer',
    paddingTop: 5
  },
  restrictedNoteLabel: {
    color: 'red',
    fontWeight: 'bold',
    paddingTop: 5
  }
};