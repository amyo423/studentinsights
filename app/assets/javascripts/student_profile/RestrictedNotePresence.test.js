import React from 'react';
import ReactDOM from 'react-dom';
import ReactTestUtils from 'react-addons-test-utils';
import renderer from 'react-test-renderer';
import fetchMock from 'fetch-mock/es5/client';
import RestrictedNotePresence from './RestrictedNotePresence';


export function testProps(props = {}) {
  return {
    eventNoteId: 42,
    studentFirstName: 'Cassandra',
    educatorName: 'Darnell',
    allowViewing: false,
    ...props
  };
}

export function mockFetch() {
  fetchMock.restore();
  fetchMock.get('express:/api/event_notes/:id/restricted_note_json', {
    text: 'RESTRICTED-text',
    event_note_attachments: [{url: 'https://website.com/RESTRICTED-url'}]
  });
}

function testEl(props = {}) {
  return <RestrictedNotePresence {...props} />;
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(testEl(props), el);
  expect(el.innerHTML).not.toContain('RESTRICTED-text');
});

it('allows viewing', done => {
  mockFetch();
  const props = testProps({allowViewing: true});
  const el = document.createElement('div');
  ReactDOM.render(testEl(props), el);
  expect(el.innerHTML).toContain('show restricted note');

  ReactTestUtils.Simulate.click($(el).find('a').get(0));
  setTimeout(() => {
    expect(el.innerHTML).not.toContain('show restricted note');
    expect(el.innerHTML).toContain('RESTRICTED-text');
    expect(el.innerHTML).toContain('This is a restricted note');
    done();
  }, 0);
});

it('snapshots', () => {
  const props = testProps();
  const tree = renderer
    .create(testEl(props))
    .toJSON();
  expect(tree).toMatchSnapshot();
});
