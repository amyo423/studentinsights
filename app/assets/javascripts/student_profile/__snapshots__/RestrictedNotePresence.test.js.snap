// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="RestrictedNotePresence"
>
  <div>
    <div
      className="NoteText"
      style={
        Object {
          "border": "1px solid transparent",
          "color": "#999",
          "fontFamily": "'Open Sans', sans-serif",
          "fontSize": 14,
          "marginTop": 5,
          "overflowX": "hidden",
          "padding": 0,
          "whiteSpace": "pre-wrap",
          "wordBreak": "break-word",
        }
      }
    >
      To respect Cassandra's privacy, Darnell marked this note as restricted.
    </div>
  </div>
</div>
`;
